data/collector.go

```diff
@@ -285,10 +338,60 @@ func (c *collector) writeCaptureResults() {
 				c.logger.Error(errors.Wrap(err, fmt.Sprintf("failed to write to collector %s", c.target.Path())).Error())
 				return
 			}
+
+			c.maybeWriteToMongo(msg)
 		}
 	}
 }
 
+// TabularData is a denormalized sensor reading.
```

Comment:
So I know what data denormalization is, but what do you mean when you say this sensor reading is denormalized?

Reply:
Each sensor reading returns this:
https://github.com/viamrobotics/api/blob/d825c9e95d385f89f951b24d0cb6f19a480905f5/proto/viam/app/datasync/v1/data_sync.proto#L4
```
// SensorMetadata contains the time the sensor data was requested and was received.
message SensorMetadata {
  google.protobuf.Timestamp time_requested = 1;
  google.protobuf.Timestamp time_received = 2;
}

// SensorData contains the contents and metadata for tabular data.
message SensorData {
  SensorMetadata metadata = 1;
  oneof data {
    google.protobuf.Struct struct = 2;
    bytes binary = 3;
  }
}
```

The collector has metadata which only changes when the collector config which is composed of:
https://github.com/viamrobotics/api/blob/d825c9e95d385f89f951b24d0cb6f19a480905f5/proto/viam/app/datasync/v1/data_sync.proto#L95
```
message UploadMetadata {
  string part_id = 1;
  string component_type = 2;
  string component_name = 3;
  string method_name = 5;
  DataType type = 6;
  string file_name = 7;
  map<string, google.protobuf.Any> method_parameters = 8;
  string file_extension = 9;
  repeated string tags = 10;

  reserved 4, 11;
  reserved "component_model", "session_id";
}
```
This means that the contents of the sensorData.readings collection isdenormalized because each document is comopnsed of the (unique) `SensorData` struct merged with the `UploadMetadata` for that reading (which is duplicated accross multiple `SensorData` structs).

If we were storing this data in a normalized fashion we would store `UploadMetadata` in a separate collection (only when it chages) and `SensorData` in it's own collection with a reference to the relevant `UploadMetadata` for that reading.

Reply:
I see, so denormalized relative to our proto API, that makes sense. Thanks!